source/MaterialXRuntime/RtFileIo.cpp

```diff
@@ -124,22 +165,77 @@ namespace
                 readCustomAttributes(elem, input, portdefAttrs);
             }
         }
+
+        return nodedefH;
     }
 
-    PvtNode* readNode(const NodePtr& src, PvtElement* parent, PvtStage* stage)
+    bool matchingSignature(const PvtNodeDef* nodedef, const RtToken nodeType, const vector<ValueElementPtr>& nodePorts)
     {
-        NodeDefPtr srcNodedef = src->getNodeDef();
-        if (!srcNodedef)
+        // Check output types for single output nodes.
+        if (nodedef->numOutputs() == 1 && nodedef->getOutput(0)->getType() != nodeType)
         {
-            throw ExceptionRuntimeError("No matching nodedef was found for node '" + src->getName() + "'");
+            return false;
+        }
+        // Check all other ports.
+        for (const ValueElementPtr& nodePort : nodePorts)
+        {
+            const PvtPortDef* nodedefPort = nodedef->findPort(RtToken(nodePort->getName()));
+            if (!nodedefPort || nodedefPort->getType().str() != nodePort->getType())
+            {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    PvtDataHandle resolveNodeDef(const NodePtr& node, PvtStage* stage)
```

Comment:
This new logic will extend the search for nodedefs to include the nodedefs already loaded in the runtime stage. This is a big improvement as it means all nodedefs no longer need to be included in the MaterialX document in order for a file to be loaded. Libraries with nodedefs can now be pre-loaded and referenced before loading MaterialX files with content data.